### PR TITLE
[IITO] Agents can shared intended common pool donations

### PR DESCRIPTION
**Pull request by Eirikalb**
# Summary

Agents can share intended common pool donations in a new IITO session
## Additional Information

Auxiliary information goes here.

## Test Plan

- Test the distribution of the contribution intentions

### CODE REVIEW

File: internal/common/baseclient/iito.go [hunk 1]
```diff
@@ -76,6 +78,35 @@ func (c *BaseClient) ReceivedGift(received shared.Resources, from shared.ClientI
 	// myResources := c.ServerReadHandle.GetGameState().ClientInfo.Resources
 }
 
+// ShareIntendedContributionis called on each client to give them the
```
Comment (lhl2617) [inline internal/common/baseclient/iito.go]:
```suggestion
// ShareIntendedContribution is called on each client to give them the
```

File: internal/server/iito.go [hunk 2]
```diff
@@ -247,3 +250,48 @@ func (s *SOMASServer) executeTransactions(transactions map[shared.ClientID]share
 		}
 	}
 }
+
+func (s *SOMASServer) runIntendedContributionSession() {
+	s.logf("start runIntendedContributionSession")
+	defer s.logf("finish runIntendedContributionSession")
+	islandContributionDict := s.getIntendedContribution()
+
```
Comment (Darrekt) [inline internal/server/iito.go]: No need for this newline (petty nitpick)

File: internal/server/iito.go [hunk 3]
```diff
@@ -247,3 +250,48 @@ func (s *SOMASServer) executeTransactions(transactions map[shared.ClientID]share
...
+	s.distributeIntendedContributions(islandContributionDict)
+}
+
+func (s *SOMASServer) getIntendedContribution() shared.IntendedContributionDict {
+	islandPredictionsDict := shared.IntendedContributionDict{}
+	nonDeadClients := getNonDeadClientIDs(s.gameState.ClientInfos)
+	for _, id := range nonDeadClients {
```
Comment (Darrekt) [inline internal/server/iito.go]:
```suggestion
	for _, id := range getNonDeadClientIDs(s.gameState.ClientInfos) {
```

File: internal/server/iito.go [hunk 4]
```diff
@@ -247,3 +250,48 @@ func (s *SOMASServer) executeTransactions(transactions map[shared.ClientID]share
...
+		c := s.clientMap[id]
+		islandPredictionsDict[id] = c.ShareIntendedContribution()
```
Comment (Darrekt) [inline internal/server/iito.go]:
```suggestion
		islandPredictionsDict[id] = s.clientMap[id].ShareIntendedContribution()
```

**Review by Darrekt — APPROVED** [review]
Thank you for writing a test with your code <3
Some nitpicky refactors, but otherwise approved.

File: internal/common/shared/contribution.go [hunk 5]
```diff
@@ -0,0 +1,21 @@
+package shared
+
+// IntendedContribution is a struct containing the intended contribution to the common pool 
+// of an island
+type IntendedContribution struct {
+	Contribution float64
```
Comment (Darrekt) [inline internal/common/shared/contribution.go]: Should be `Resource` instead of `float64`

File: internal/common/shared/contribution.go [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+	TeamsOfferedTo []ClientID
+}
+
+// ReceivedIntendedContribution is a struct containing the information describing 
+// an intended contribution made by a given island
+type ReceivedIntendedContribution struct {
+	Contribution float64
```
Comment (Darrekt) [inline internal/common/shared/contribution.go]: Same here, needs to change throughout the PR.

**Review by Darrekt — DISMISSED** [review]
Sorry, missed it during my first review. Please change all `float64`'s representing contributions to `Resources`

File: internal/common/baseclient/iito.go [hunk 7]
```diff
@@ -76,6 +78,35 @@ func (c *BaseClient) ReceivedGift(received shared.Resources, from shared.ClientI
...
+// option of sharing how much they intend to contribute to the common pool. They must also
+// choose what islands they want to share the information with
+func (c *BaseClient) ShareIntendedContribution() shared.IntendedContribution {
```
Comment (E-Mac27) [inline internal/common/baseclient/iito.go]: Please add Compulsory/Optional

File: internal/server/iito_test.go [hunk 8]
```diff
@@ -303,3 +334,59 @@ func TestDistributeGiftHistory(t *testing.T) {
...
+	server.distributeIntendedContributions(input)
+	for id := range mockClient {
+		got := mockClient[id].getOtherIslandsCommonPoolContribution()
+		w := want[id]
+		if !reflect.DeepEqual(w, got) {
+			t.Errorf("want '%#v' got '%#v' for %#v", w, got, id)
+		}
+	}
+
+}
```
Comment (Darrekt) [inline internal/server/iito_test.go]: image

**Review by E-Mac27 — COMMENTED** [review]
👍 Nice.

**Comment by Eirikalb** [comment]
> Sorry, missed it during my first review. Please change all `float64`'s representing contributions to `Resources`

Done this now

File: internal/common/baseclient/iito.go [hunk 9]
```diff
@@ -76,6 +78,36 @@ func (c *BaseClient) ReceivedGift(received shared.Resources, from shared.ClientI
...
 
+// ShareIntendedContribution is called on each client to give them the
+// option of sharing how much they intend to contribute to the common pool. They must also
+// choose what islands they want to share the information with
+// OPTIONAL, you can implement this if you want to partake in the iito session
+func (c *BaseClient) ShareIntendedContribution() shared.IntendedContribution {
+
+	// For MVP, share this prediction with all islands since trust has not yet been implemented
+	trustedIslands := make([]shared.ClientID, len(RegisteredClients))
+	for index, id := range shared.TeamIDs {
```
Comment (NGovani) [inline internal/common/baseclient/iito.go:89]: Would be better to use the nonDeadAgents functionality through the ServerReadHandle

File: internal/common/baseclient/iito.go [hunk 10]
```diff
@@ -76,6 +78,36 @@ func (c *BaseClient) ReceivedGift(received shared.Resources, from shared.ClientI
...
+	}
+
+	c.intendedContribution = contribution
+	return contribution
+
+}
+
+// ReceiveIntendedContribution provides each client with the intended common pool contribution from the islands
+// that have chosen to share it with them
+// COMPULSORY, you must implement this function
```
Comment (NGovani) [inline internal/common/baseclient/iito.go]: should be switched to Optional, or the other function should be made compulsory. I'd rather both be compulsary

**Review by NGovani — APPROVED** [review]
Some issues with the base client but the infra stuff looks good enough to merge

File: internal/common/baseclient/iito.go (hunk 9 above)
Comment (Eirikalb) [reply internal/common/baseclient/iito.go:89]: Agreed

File: internal/common/baseclient/iito.go (hunk 10 above)
Comment (Eirikalb) [reply internal/common/baseclient/iito.go]: Agreed

**Pull request merged**